client/client.go

```diff
@@ -1115,91 +1113,107 @@ func (r *NotaryRepository) bootstrapClient(checkInitialized bool) (*TUFClient, e
+// managing the key to the server. If key(s) are specified by keyList, then they are
+// used for signing the role.
+// These changes are staged in a changelist until publish is called.
+func (r *NotaryRepository) RotateKey(role data.RoleName, serverManagesKey bool, keyList []string) error {
+	return rotateKey(r, role, serverManagesKey, keyList, nil)
+}
+
+// RotateKeyWithCert removes all existing keys associated with the role.
+// If no certificates are specified in certs then throw an error.
+// the key and certificate pairs are used for signing the role.
+// These changes are staged in a changelist until publish is called.
+func (r *NotaryRepository) RotateKeyWithCert(role data.RoleName, serverManagesKey bool, keyList []string, certs []data.PublicKey) error {
+	return rotateKey(r, role, serverManagesKey, keyList, certs)
+}
+
+// pubKeyListForRotation rotates keys associated with role using the provided pub-priv key pairs.
+// Returns the public keys used in the rotation.
+// Create new key pair if neither private nor public key is provided.
+func (r *NotaryRepository) pubKeyListForRotation(role data.RoleName, serverManaged bool, privKeys []string, pubKeys data.KeyList) (pubKeyList data.KeyList, err error) {
 	var pubKey data.PublicKey
 
-	// If server manages the key being rotated, request a rotation and return the new key
-	if serverManaged {
+	// for each of the following 5 cases:
+	// validate and generate key pairs if necessary
+	switch {
+	case serverManaged:
+		// If server manages the key being rotated, request a rotation and return the new key
 		remote := r.getRemoteStore()
 		pubKey, err = rotateRemoteKey(role, remote)
-		pubKeyList = make(data.KeyList, 0, 1)
 		pubKeyList = append(pubKeyList, pubKey)
-		if err != nil {
-			return nil, fmt.Errorf("unable to rotate remote key: %s", err)
+		return pubKeyList, err
+	case len(pubKeys) == 0 && len(privKeys) == 0:
+		if pubKey, err = r.CryptoService.Create(role, r.gun, data.ECDSAKey); err != nil {
+			return nil, fmt.Errorf("unable to generate key: %s", err)
 		}
-		return pubKeyList, nil
-	}
-
-	// If no new keys are passed in, we generate one
-	if len(newKeys) == 0 {
-		pubKeyList = make(data.KeyList, 0, 1)
-		pubKey, err = r.CryptoService.Create(role, r.gun, data.ECDSAKey)
 		pubKeyList = append(pubKeyList, pubKey)
-	}
-	if err != nil {
-		return nil, fmt.Errorf("unable to generate key: %s", err)
-	}
-
-	// If a list of keys to rotate to are provided, we add those
-	if len(newKeys) > 0 {
-		pubKeyList = make(data.KeyList, 0, len(newKeys))
-		for _, keyID := range newKeys {
-			pubKey = r.CryptoService.GetKey(keyID)
-			if pubKey == nil {
+	case len(pubKeys) == 0 && len(privKeys) != 0:
+		for _, keyID := range privKeys {
+			if pubKey = r.CryptoService.GetKey(keyID); pubKey == nil {
 				return nil, fmt.Errorf("unable to find key: %s", keyID)
 			}
 			pubKeyList = append(pubKeyList, pubKey)
 		}
+	case len(pubKeys) != 0 && len(privKeys) != 0:
```

Comment:
Thanks David that makes sense. I will make an update to allow more public keys than private keys and also loose the restriction on ordering keys. I have kept the 4 cases distinct because it might be easier to understand?? Let me know what you think!